cvmfs/swissknife_filestats.cc

```diff
@@ -0,0 +1,317 @@
+    (args.count('k') > 0) ? *args.find('k')->second : "";
+  if (DirectoryExists(repo_keys))
+    repo_keys = JoinStrings(FindFilesBySuffix(repo_keys, ".pub"), ":");
+  const std::string &tmp_dir   =
+    (args.count('l') > 0) ? *args.find('l')->second : "/tmp";
+  if (args.count('h') > 0) {
+    manual_root_hash = shash::MkFromHexPtr(shash::HexPtr(
+      *args.find('h')->second), shash::kSuffixCatalog);
+  }
+
+  tmp_db_path_ = "/tmp/cvmfs_filestats/";
+  num_downloaded_ = 0;
+
+  bool success = false;
+  if (IsHttpUrl(repo_url)) {
+    const bool follow_redirects = false;
+    if (!this->InitDownloadManager(follow_redirects) ||
+        !this->InitVerifyingSignatureManager(repo_keys)) {
+      LogCvmfs(kLogCatalog, kLogStderr, "Failed to init remote connection");
+      return 1;
+    }
+
+    HttpObjectFetcher<catalog::Catalog,
+                      history::SqliteHistory> fetcher(repo_name,
+                                                      repo_url,
+                                                      tmp_dir,
+                                                      download_manager(),
+                                                      signature_manager());
+    success = Run(&fetcher);
+  } else {
+    LocalObjectFetcher<> fetcher(repo_url, tmp_dir);
+    success = Run(&fetcher);
+  }
+
+  return (success) ? 0 : 1;
+}
+
+template <class ObjectFetcherT>
+bool CommandFileStats::Run(ObjectFetcherT *object_fetcher)
+{
+  running_ = true;
+
+  string abs_path = GetAbsolutePath(db_path_);
+  unlink(abs_path.c_str());
+  db_ = FileStatsDatabase::Create(db_path_);
+  db_->InitStatements();
+
+  RemoveTree(tmp_db_path_);
+  assert(MkdirDeep(tmp_db_path_, 0755));
+
+  typename CatalogTraversal<ObjectFetcherT>::Parameters params;
+  params.object_fetcher = object_fetcher;
+  CatalogTraversal<ObjectFetcherT> traversal(params);
+  traversal.RegisterListener(&CommandFileStats::CatalogCallback, this);
+
+  pthread_create(&thread_processing_, NULL, MainProcessing, this);
+
+  bool ret = traversal.Traverse();
+
+  running_ = false;
+  pthread_join(thread_processing_, NULL);
+
+  db_->DestroyStatements();
+
+  return ret;
+}
+
+void CommandFileStats::CatalogCallback(
+  const CatalogTraversalData<catalog::Catalog> &data) {
+  string out_path =  tmp_db_path_ + StringifyInt(num_downloaded_ + 1) + ".db";
+  assert(CopyPath2Path(data.catalog->database_path(), out_path));
+  ++num_downloaded_;
+}
+
+void *CommandFileStats::MainProcessing(void *data) {
+  CommandFileStats *repo_stats = static_cast<CommandFileStats *>(data);
+  int processed = 0;
+  while (repo_stats->running_ || processed < repo_stats->num_downloaded_) {
+    if (processed < repo_stats->num_downloaded_) {
+      LogCvmfs(kLogCatalog, kLogStdout, "Processing catalog %d", processed);
+      string db_path = repo_stats->tmp_db_path_ + "/" +
+                       StringifyInt(processed + 1) + ".db";
+      repo_stats->ProcessCatalog(db_path);
+      unlink(db_path.c_str());
```

Comment:
The cvmfs sqlite database base class has the concept of file ownership.  Can this be used to avoid the manual `unlink()`?

Reply:
Thanks for the advice. Manually taking file ownership after opening the database deletes the db file upon database object destruction.